numba/np/ufunc/gufunc.py

```diff
@@ -0,0 +1,116 @@
+from numba import typeof
+from numba.core import types
+from numba.np.ufunc.ufuncbuilder import GUFuncBuilder
+from numba.np.ufunc.sigparse import parse_signature
+
+
+class GUFunc(object):
+    """
+    Dynamic generalized universal function (GUFunc)
+    intended to act like a normal Numpy gufunc, but capable
+    of call-time (just-in-time) compilation of fast loops
+    specialized to inputs.
+    """
+
+    def __init__(self, py_func, signature, identity=None, cache=None,
+                 targetoptions={}):
+        self.ufunc = None
+        self.signature = signature
+        self.gufunc_builder = GUFuncBuilder(
+            py_func, signature, identity, cache, targetoptions)
+
+    def add(self, fty):
+        self.gufunc_builder.add(fty)
+
+    def build_ufunc(self):
+        self.ufunc = self.gufunc_builder.build_ufunc()
+        return self
+
+    @property
+    def __doc__(self):
+        return self.ufunc.__doc__
+
+    @property
+    def __name__(self):
+        return self.ufunc.__name__
+
+    @property
+    def nin(self):
+        return self.ufunc.nin
+
+    @property
+    def nout(self):
+        return self.ufunc.nout
+
+    @property
+    def nargs(self):
+        return self.ufunc.nargs
+
+    @property
+    def ntypes(self):
+        return self.ufunc.ntypes
+
+    @property
+    def types(self):
+        return self.ufunc.types
+
+    @property
+    def identity(self):
+        return self.ufunc.identity
+
+    def _get_argtys_dtype(self, args):
+        argtys = map(lambda x: typeof(x), args)
+        tys = []
+        for argty in argtys:
+            if isinstance(argty, types.Array):
+                tys.append(argty.dtype)
+            else:
+                tys.append(argty)
+        return tys
+
+    def _get_element_wise_signature(self, *args):
+        parsed_sig = parse_signature(self.signature)
+        dtypes = self._get_argtys_dtype(args)  # [int32, int32, int32, ...]
+
+        # Two cases here can happen here
+        # 1. args contains the output array:
+        #      gufunc(A, B, C)
+        # 2. args doesn't contains the output array:
+        #      C = gufunc(A, B)
+        #
+        # In the latter case, one would have to guess the type
+        # and format (scalar or array). This behavior will be
+        # forbidden for now.
+        # See: https://github.com/numba/numba/pull/5938#issuecomment-661978921
+
+        # self.signature[1] has always length 1
+        if len(dtypes) < len(self.signature[0]) + 1:
+            msg = '`out = gufunc(Arg1, Arg2, ..., ArgN)` is not allowed.'
+            '\nUse `gufunc(Arg1, Arg2, ..., ArgN, out)` instead'
+            raise TypeError(msg)
```

Comment:
Is it possible for the number of args to be wrong for other reasons (like forgetting an argument). Hints like this can be nice but they can lead the user astray when they don't actually match what happened. Maybe at least prefix the message with "too few arguments" or something like that.